Elaborate on the differences between dev mode and continuous builds

thomasheartman opened the pull request:
This PR adds some extra details to the descriptions of dev mode and
continuous builds. In particular, it:
- elaborates on what dev mode does, saying that it also picks up
  changes to XML schemas
- specifies what sort of tasks you need continuous builds for

As always, any feedback is much appreciated, especially in regard to
factual inaccuracies and the like.

thomasheartman commented:
@rymsha I'm moving the discussion on the change set here. Hope you don't mind. 

Could you tell me a little bit more about what changes are made to the caching in dev mode? Also, do the changes introduced in 36238cb correct the mistakes on dev mode file sourcing and warn the user about dev mode in prod?

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,42 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
+Enonic XP can be run in a special mode known as _development mode_. In this mode, XP will detect any changes to your project source files, including JavaScript controllers and XML schemas. It does this by reading these files directly from their source locations instead of from the application `.jar` file.
 
-You may activate dev mode when starting XP as follows:
+Development mode also disables some of XP's caching mechanisms. (What mechanisms and why?)
```
- assetUrl function generates new fingerprint for every request, so browsers can't cache assets
- resource cache invalidated if resource `modified time`changes.
- entire JavaScript `require` cache invalidated  if any of required resources  `modified time`changes.

thomasheartman commented:
@rymsha  I spoke to @sigdestad and he also clarified for me that XP doesn't pick up *new* content types when running in dev mode, so I added that to the section. Do you know whether this applies only to content types or whether it's *any* XML schema?

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,51 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
-You may activate dev mode when starting XP as follows:
+[NOTE]
+====
+Because of the way that XP works, some changes are not picked up in dev mode: if you add a _new_ content type, XP will not pick it up automatically. This is because content types are registered with XP when _deploying_ an app. However, changes to existing content types _are_ detected. If you'd like to have new content types picked up automatically, try using xref:build-system#_continuous_building ...
```
> This is because content types are registered with XP when _deploying_ an app

I'd rather take a look at it. Application loader code is complicated and there might be a bug, but I can't confirm that content types get registered when app is deployed.

thomasheartman replied:
Aight, got it. I'll look into rewording this next week. It's not really important *why* it doesn't work, as long as the reader knows that it doesn't. Thanks!

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,51 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
+====
+Because of the way that XP works, some changes are not picked up in dev mode: if you add a _new_ content type, XP will not pick it up automatically. This is because content types are registered with XP when _deploying_ an app. However, changes to existing content types _are_ detected. If you'd like to have new content types picked up automatically, try using xref:build-system#_continuous_building ...
+
+This does not affect JavaScript controllers. You can add new controllers and they will be picked up by XP automatically.
```
This is where things get complicated for dev mode:
if controllers are written in ES6 or TypeScript they require transpiling. Only vanilla JS controllers can work without `gradle build` step.
Also applies for css vs sass/less

rymsha replied:
You mention it later, so probably it is irrelevant.

thomasheartman replied:
Why does ES6 require transpiling? From what I understand, JS controllers are somehow interpreted by the Java runtime (using Nashorn?), and this doesn't support the full ES6 feature set. Is this why? 

Not to be pedantic, but I think this is a good distinction to make: when you say "vanilla JS", you mean ES5, is this correct? I think it would be misleading to use "vanilla JS" to refer to pre-ES6 JS in the docs, considering ES6 (and later versions) are still "vanilla".

sigdestad replied:
Nashorn supports ES6 partly, there are overview tables that explain what it supports and not. Perhaps we should link to that?

thomasheartman replied:
That might be useful. Is this documented elsewhere in the XP docs (or other Enonic docs)? I don't think we need to go into intricate detail here, but I think would be helpful to say something along the lines of 

> JavaScript controllers that require some sort of compilation step won't be sourced directly. To find out more about what features are and are not available, check out <these docs>

thomasheartman replied:
@sigdestad Are these nashorn ES6 compatibility tables available anywhere? I found this [JEP](https://openjdk.java.net/jeps/292), but it doesn't have a clear overview listed.

sigdestad replied:
I've used this one: https://kangax.github.io/compat-table/es6/
Nashorn is called JJS afaik.

thomasheartman commented:
Alright, I've added links to Kangax' compatibility table and updated various parts of the doc now according to feedback. Is there anything else we'd like to change/fix before this is ready to go? :relaxed:

thomasheartman commented:
@sigdestad @rymsha Any more thoughts? ☝🏾

rymsha commented:
I'm pretty sure, Typescript "compiled" into vanilla JS files also picked up. It is just less useful - because after build step it is almost-instant deploy step.
So either we remove lookup for the files in build folder, or we change this documentation to say that files are picked up after build step, too.

thomasheartman commented:
@sry Thanks for the reply and sorry for taking so long to get back to this.

> I'm pretty sure, Typescript "compiled" into vanilla JS files also picked up. It is just less useful - because after build step it is almost-instant deploy step.
> So either we remove lookup for the files in build folder, or we change this documentation to say that files are picked up after build step, too.

I have a few questions regarding what you're saying:

You say that files are deployed after the build step. Is there a build step if XP runs in dev mode? From what I understand, XP running in dev mode won't actually build anything, it'll just source files directly from their locations on disk. I thought that if you had a build step (such as compiling typescript), you'd need to set up a continuous build instead with a custom build step?

So, if my above assumption is correct, then XP will probably pick up changes to the TS files, but because XP can't work with TS directly, it doesn't do anything with them. Is that correct or is there something else I'm missing here?

Or maybe, when using typescript or other compile-to-js solutions, it's common to have a separate build process running (such as webpack, esbuild) and that's what you're talking about?

rymsha commented:
> Is there a build step if XP runs in dev mode?

Build is not in XP itself. Build is done by Gradle (`./gradlew build`). And it may contain a step to convert .ts file into .js files and put them in "build" folder.
Next step in `gradle build` is to put .js files from build folder into "jar" archive (also somewhere in build folder)
That "jar" is copied to XP deploy folder by `./gradlew deploy`

For other questions I have no answer for now. I need to investigate https://github.com/enonic/xp/issues/8835

thomasheartman commented:
Got it! 

Then, unless you explicitly run a build process next to running XP in dev mode, files that need to be processed (compiled or otherwise), will not automatically update, right? I have updated the text to reflect that, I think.

thomasheartman commented:
@rymsha  do you have further comments after the latest changes? I feel like this PR has been open long enough (though that's on me!) 😅

rymsha commented:
We are planning to discuss dev mode some time soon, but for now this PR has to remain open. Sorry.

thomasheartman commented:
Aight; no problem. Just let me know when you're ready to pick it back up ☺️

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,54 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
+
+- Files that require any sort of compilation (including transpilation) are not detected unless you run a build process on the side. For instance, if you use TypeScript or JavaScript features that XP doesn't support, these files won't be picked up by default. To have XP detect them, run a separate build process on the side, such as either a xref:build-system#_continuous_building[continuous build]  ...
++
+NOTE: XP uses the  Nashorn JavaScript engine for compiling JavaScript. For an overview over what JavaScript features Nashorn supports, see https://kangax.github.io/compat-table/es6/#nashorn1_8[this feature table].
```
```suggestion
NOTE: XP uses the  Nashorn JavaScript engine for executing JavaScript. For an overview over what JavaScript features Nashorn supports, see https://kangax.github.io/compat-table/es6/#nashorn1_8[this feature table].
```

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,54 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
++
+NOTE: XP uses the  Nashorn JavaScript engine for compiling JavaScript. For an overview over what JavaScript features Nashorn supports, see https://kangax.github.io/compat-table/es6/#nashorn1_8[this feature table].
++
+- XP will not pick up _new_ content types. If you add a new content type after your previous project deploy, XP will not be able to see it. However, changes to _existing_ content types _are_ picked up. XP will pick up new JavaScript controllers as long as it can run them without compilation (as per the previous point).
```
> XP will not pick up _new_ content types
Is it true @anatol-sialitski ?

anatol-sialitski replied:
I've just test that and as the result a new content type is appeared in Content Studio without redeploy a project. Pay attention that I test that after [this](https://github.com/enonic/xp-gradle-plugin/pull/83) fix in gradle plugin.

rymsha replied:
```suggestion
```

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,54 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
+
+Development mode also disables some of XP's caching mechanisms. To make the development workflow as smooth as possible, XP tries to invalidate caches for your static assets. This is to prevent you from getting stale resources so that what you see in the browser is always as up to date as possible.
+
+Running XP in development mode may be convenient if your project does not rely on additional build steps such as JavaScript bundling. If you need to run arbitrary build tasks before deploying your project, check out xref:build-system#_continuous_building[the section on continuous builds].
```
This is confusing... webpack recreates the bundle in build directory and new bundle sis picked up nicely.MAybe remove this paragraph alltogether?

alansemenov replied:
```suggestion
```

rymsha commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,54 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
++
+- XP will not pick up _new_ content types. If you add a new content type after your previous project deploy, XP will not be able to see it. However, changes to _existing_ content types _are_ picked up. XP will pick up new JavaScript controllers as long as it can run them without compilation (as per the previous point).
+
+If you need any of the above features, try using xref:build-system#_continuous_building[continuous build mode] instead.
```
```suggestion
If development mode does not work for you, try using xref:build-system#_continuous_building[continuous build mode] instead.
```

thomasheartman commented:
Hey! Just a heads-up: I can see the changes you've suggested and I support them, but I can't accept them because I don't have write-access to this repo. I suggest you take over this PR and take it from here instead. Sound alright? ☺️

alansemenov commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,53 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
 
 Gradle also supports a https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continuous_build[continuous build mode].
-This will monitor your project assets for changes and run the specified task when something changes.
+In continuous build mode, Gradle will monitor your project assets for changes and run a task you specify task when something changes.
```
```suggestion
In continuous build mode, Gradle will monitor your project assets for changes and run a task you specify when something changes.
```

alansemenov commented on docs/apps/build-system.adoc:
```diff
@@ -89,33 +89,53 @@ NOTE:  XP apps deployed via file (locally) are presented with a small blue icon
+
+Continuous builds come in handy when the changes you're working on require a full compile and redeploy, such as when you're working with Java or need to build/compile client-side assets with an external build tool (such as webpack).
+
+Another use case for continuous builds is if you're adding _new_ content types (or other XML schemas?). Because these are registered when an app is deployed, you need to compile and redeploy your app for XP to become aware of them.
```
```suggestion
```
